### ISSUE TITLE
[RLlib] ParametricRecSys observations  outside of observation space

### ISSUE THREAD

**Issue by vincent-lous** [ray-project/ray#28231]
### What happened + What you expected to happen

I want to evaluate a PPOTrainer that I ran on a ray.rllib.examples.env.bandit_envs_recommender_system.ParametricRecSys-environment. This gives an error that the observation is outside given space. 

> "name": "ValueError",
> 	"message": "('Observation ({} dtype={}) outside given space ({})!', {'user': array([ 0.54109985, -0.55572975,  0.59655416, -0.9165882 ,  0.32756427,\n       -0.8653462 , -0.61950433, -0.16755551,  0.14229618, -0.5263172 ,\n       -0.35305858,  0.4152433 , -0.34991303,  0.8304454 , -0.941373  ,\n       -0.5386489 , -0.5765393 ,  0.93252766,  0.19801164,  0.20225774],\n      dtype=float32), 'doc': {'0': array([ 0.5215836 , -0.61120254, -0.35855845,  0.08027998, -0.01534893,\n       -0.30238396,  0.52114433, -0.40023103, -0.9608099 , -0.84224933,\n        0.18388008,  0.8929905 , -0.94119525, -0.27208805,  0.04744889,\n        0.18770689,  0.06966066, -0.8296684 ,  0.17139104,  0.9026296 ],\n      dtype=float32), '1': array([ 0.40011796,  0.90483886,  0.7404875 , -0.3168922 ,  0.7561373 ,\n        0.44064188, -0.41803527, -0.71604526,  0.7533913 , -0.66952837,\n       -0.8887174 , -0.25062314, -0.72037303, -0.7586206 , -0.44427392,\n       -0.2057975 ,  0.9175719 ,  0.08363984,  0.41493717,  0.73632693],\n      dtype=float32), '2': array([-0.3879901 ,  0.8895779 , -0.53160536,  0.99060726, -0.25774595,\n       -0.10323463, -0.24117081,  0.19701774,  0.40639964, -0.1330463 ,\n       -0.5176396 , -0.80891657,  0.8413943 , -0.64483243, -0.17267403,\n        0.28457406, -0.9670741 ,  0.108579  ,  0.6812572 ,  0.8509619 ],\n      dtype=float32), '3': array([ 0.12664089, -0.6939673 , -0.8273438 ,  0.52983147, -0.24593566,\n        0.47093043, -0.48062876,  0.74770284, -0.49259254, -0.01618658,\n       -0.728846  ,  0.6149716 , -0.23746905, -0.3409355 ,  0.77106273,\n       -0.44993547, -0.9026514 ,  0.9584925 , -0.4284208 ,  0.8295323 ],\n      dtype=float32), '4': array([-0.53013366,  0.6375033 ,  0.8533416 ,  0.08790456, -0.5630317 ,\n       -0.30186448, -0.09573772, -0.930947  ,  0.1371054 ,  0.03762012,\n        0.19135992, -0.4111602 ,  0.40417612,  0.30070683,  0.30006787,\n       -0.4560212 ,  0.6658053 , -0.5562488 ,  0.4876351 ,  0.36921647],\n      dtype=float32), '5': array([ 0.8954167 , -0.656551  ,  0.03998402,  0.3623367 , -0.5852495 ,\n       -0.90839344,  0.78480273, -0.9067722 ,  0.46540296,  0.9860009 ,\n        0.85866815,  0.77937496, -0.88849443,  0.6600391 ,  0.20394546,\n       -0.40542805, -0.63944346,  0.5085564 ,  0.41492102,  0.7786271 ],\n      dtype=float32), '6': array([ 0.81958085, -0.58244085, -0.55936784,  0.667276  , -0.01108886,\n       -0.091013  , -0.06516552,  0.12307017,  0.97585344,  0.5506294 ,\n        0.93281275,  0.16625397,  0.3985712 ,  0.7806259 ,  0.80137193,\n       -0.05712234, -0.9431896 , -0.8236043 ,  0.23951314, -0.08919021],\n      dtype=float32), '7': array([-0.8161121 , -0.7021542 ,  0.65553284,  0.0316354 ,  0.96013194,\n        0.68845767, -0.19276679,  0.44326198,  0.67540234, -0.6394616 ,\n        0.79296744,  0.8019955 , -0.02458191,  0.18718313, -0.06069673,\n       -0.20879127,  0.7138461 , -0.11197664, -0.71673816, -0.21307509],\n      dtype=float32), '8': array([-0.17533064, -0.7681836 ,  0.01002871, -0.6397116 , -0.827161  ,\n       -0.41476434, -0.3089181 ,  0.26077452, -0.6416258 ,  0.4931263 ,\n       -0.05076075, -0.6839608 , -0.6767113 ,  0.6149199 ,  0.6266382 ,\n       -0.64473087,  0.31474325, -0.28244218,  0.5364549 ,  0.6604383 ],\n      dtype=float32), '9': array([-8.9509588e-01,  4.4338930e-01,  4.9490455e-01,  2.7083248e-01,\n        5.4459065e-01,  7.0892900e-01,  6.3318557e-01,  5.2825874e-01,\n        5.2950054e-04,  1.3722055e-01, -5.5846149e-01,  4.4012126e-01,\n       -7.1953773e-01,  9.4238293e-01,  5.4676449e-01, -4.2552534e-01,\n       -7.8317828e-02, -8.0405241e-01, -9.9607080e-01, -2.5939114e-02],\n      dtype=float32)}, 'response': ({'click': 1, 'engagement': array(-0.39011952, dtype=float32)},)}, None, Dict(doc:Dict(0:Box([-1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1.\n -1. -1.], [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.], (20,), float32), 1:Box([-1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1.\n -1. -1.], [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.], (20,), float32), 2:Box([-1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1.\n -1. -1.], [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.], (20,), float32), 3:Box([-1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1.\n -1. -1.], [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.], (20,), float32), 4:Box([-1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1.\n -1. -1.], [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.], (20,), float32), 5:Box([-1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1.\n -1. -1.], [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.], (20,), float32), 6:Box([-1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1.\n -1. -1.], [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.], (20,), float32), 7:Box([-1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1.\n -1. -1.], [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.], (20,), float32), 8:Box([-1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1.\n -1. -1.], [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.], (20,), float32), 9:Box([-1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1.\n -1. -1.], [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.], (20,), float32)), response:Tuple(Dict(click:Discrete(2), engagement:Box(0.0, 100.0, (), float32))), user:Box([-1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1. -1.\n -1. -1.], [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.], (20,), float32)))"

### Versions / Dependencies

My conda environment. 
```
Python 3.10.5
dependencies:
  - abseil-cpp=20210324.2=hbdafb3b_0
  - absl-py=1.2.0=pyhd8ed1ab_0
  - aiohttp=3.8.1=py310hf8d0d8f_1
  - aiosignal=1.2.0=pyhd8ed1ab_0
  - appnope=0.1.2=py310hca03da5_1001
  - asttokens=2.0.5=pyhd3eb1b0_0
  - astunparse=1.6.3=pyhd8ed1ab_0
  - async-timeout=4.0.2=pyhd8ed1ab_0
  - attrs=22.1.0=pyh71513ae_1
  - backcall=0.2.0=pyhd3eb1b0_0
  - blinker=1.4=py_1
  - brotlipy=0.7.0=py310hf8d0d8f_1004
  - bzip2=1.0.8=h620ffc9_4
  - c-ares=1.18.1=h3422bc3_0
  - ca-certificates=2022.07.19=hca03da5_0
  - cached-property=1.5.2=hd8ed1ab_1
  - cached_property=1.5.2=pyha770c72_1
  - cachetools=4.2.4=pyhd8ed1ab_0
  - certifi=2022.6.15=py310hca03da5_0
  - cffi=1.15.1=py310he00a5c5_0
  - charset-normalizer=2.1.0=pyhd8ed1ab_0
  - cryptography=37.0.4=py310h94bb23d_0
  - debugpy=1.5.1=py310hc377ac9_0
  - decorator=5.1.1=pyhd3eb1b0_0
  - entrypoints=0.4=py310hca03da5_0
  - executing=0.8.3=pyhd3eb1b0_0
  - frozenlist=1.3.1=py310h8c01e39_0
  - gast=0.4.0=pyh9f0ad1d_0
  - giflib=5.2.1=h27ca646_2
  - google-auth=1.35.0=pyh6c4a22f_0
  - google-auth-oauthlib=0.4.6=pyhd8ed1ab_0
  - google-pasta=0.2.0=pyh8c360ce_0
  - grpc-cpp=1.42.0=hedfbb7c_1
  - h5py=3.6.0=nompi_py310hb8bbf05_100
  - hdf5=1.12.1=nompi_had0e5e0_103
  - icu=69.1=hbdafb3b_0
  - idna=3.3=pyhd8ed1ab_0
  - importlib-metadata=4.11.4=py310hbe9552e_0
  - ipykernel=6.9.1=py310hca03da5_0
  - ipython=8.4.0=py310hca03da5_0
  - jedi=0.18.1=py310hca03da5_1
  - jpeg=9e=he4db4b2_2
  - jupyter_client=7.2.2=py310hca03da5_0
  - jupyter_core=4.10.0=py310hca03da5_0
  - keras=2.7.0=pyhd8ed1ab_0
  - keras-preprocessing=1.1.2=pyhd8ed1ab_0
  - krb5=1.19.2=hd92b7a7_3
  - libblas=3.9.0=15_osxarm64_openblas
  - libcblas=3.9.0=15_osxarm64_openblas
  - libcurl=7.81.0=h8fe1914_0
  - libcxx=12.0.0=hf6beb65_1
  - libedit=3.1.20191231=hc8eb9b7_2
  - libev=4.33=h642e427_1
  - libffi=3.4.2=hc377ac9_4
  - libgfortran=5.0.0.dev0=11_0_1_hf114ba7_23
  - libgfortran5=11.0.1.dev0=hf114ba7_23
  - liblapack=3.9.0=15_osxarm64_openblas
  - libnghttp2=1.46.0=he4cd7f6_0
  - libopenblas=0.3.20=openmp_h130de29_1
  - libpng=1.6.37=h2c9beb0_3
  - libprotobuf=3.19.4=hccf11d3_0
  - libsodium=1.0.18=h1a28f6b_0
  - libssh2=1.10.0=hb80f160_2
  - libzlib=1.2.12=ha287fd2_2
  - llvm-openmp=14.0.4=hd125106_0
  - markdown=3.4.1=pyhd8ed1ab_0
  - markupsafe=2.1.1=py310hf8d0d8f_1
  - matplotlib-inline=0.1.2=pyhd3eb1b0_2
  - multidict=6.0.2=py310hf8d0d8f_1
  - ncurses=6.3=h1a28f6b_3
  - nest-asyncio=1.5.5=py310hca03da5_0
  - oauthlib=3.2.0=pyhd8ed1ab_0
  - openssl=1.1.1q=h1a28f6b_0
  - opt_einsum=3.3.0=pyhd8ed1ab_1
  - parso=0.8.3=pyhd3eb1b0_0
  - pexpect=4.8.0=pyhd3eb1b0_3
  - pickleshare=0.7.5=pyhd3eb1b0_1003
  - pip=22.1.2=py310hca03da5_0
  - prompt-toolkit=3.0.20=pyhd3eb1b0_0
  - ptyprocess=0.7.0=pyhd3eb1b0_2
  - pure_eval=0.2.2=pyhd3eb1b0_0
  - pyasn1=0.4.8=py_0
  - pyasn1-modules=0.2.7=py_0
  - pycparser=2.21=pyhd8ed1ab_0
  - pygments=2.11.2=pyhd3eb1b0_0
  - pyjwt=2.4.0=pyhd8ed1ab_0
  - pyopenssl=22.0.0=pyhd8ed1ab_0
  - pysocks=1.7.1=py310hbe9552e_5
  - python=3.10.5=h71ab1a4_0_cpython
  - python-dateutil=2.8.2=pyhd3eb1b0_0
  - python-flatbuffers=2.0=pyhd8ed1ab_0
  - python_abi=3.10=2_cp310
  - pytorch=1.12.0=py3.10_0
  - pyu2f=0.1.5=pyhd8ed1ab_0
  - pyzmq=23.2.0=py310hc377ac9_0
  - re2=2021.11.01=hbdafb3b_0
  - readline=8.1.2=h1a28f6b_1
  - requests=2.28.1=pyhd8ed1ab_0
  - requests-oauthlib=1.3.1=pyhd8ed1ab_0
  - rsa=4.9=pyhd8ed1ab_0
  - setuptools=61.2.0=py310hca03da5_0
  - six=1.16.0=pyh6c4a22f_0
  - snappy=1.1.8=hc88da5d_3
  - sqlite=3.39.0=h1058600_0
  - stack_data=0.2.0=pyhd3eb1b0_0
  - tensorboard=2.6.0=pyhd8ed1ab_1
  - tensorboard-data-server=0.6.0=py310hf1846ab_2
  - tensorboard-plugin-wit=1.8.1=pyhd8ed1ab_0
  - tensorflow=2.7.0=cpu_py310h6ecea76_0
  - tensorflow-base=2.7.0=cpu_py310h8c911da_0
  - tensorflow-estimator=2.7.0=cpu_py310h9642b6f_0
  - termcolor=1.1.0=pyhd8ed1ab_3
  - tk=8.6.12=hb8d0fd4_0
  - tornado=6.1=py310h1a28f6b_0
  - traitlets=5.1.1=pyhd3eb1b0_0
  - typing-extensions=4.3.0=hd8ed1ab_0
  - typing_extensions=4.3.0=pyha770c72_0
  - tzdata=2022a=hda174b7_0
  - urllib3=1.26.11=pyhd8ed1ab_0
  - wcwidth=0.2.5=pyhd3eb1b0_0
  - werkzeug=2.2.1=pyhd8ed1ab_0
  - wheel=0.37.1=pyhd3eb1b0_0
  - wrapt=1.14.1=py310h02f21da_0
  - xz=5.2.5=h1a28f6b_1
  - yarl=1.7.2=py310hf8d0d8f_2
  - zeromq=4.3.4=hc377ac9_0
  - zipp=3.8.1=pyhd8ed1ab_0
  - zlib=1.2.12=h5a0b063_2
  - pip:
    - aiohttp-cors==0.7.0
    - aiorwlock==1.3.0
    - anyio==3.6.1
    - asgiref==3.5.2
    - backoff==1.10.0
    - blessed==1.19.1
    - click==8.0.4
    - cloudpickle==2.1.0
    - colorful==0.5.4
    - cycler==0.11.0
    - distlib==0.3.5
    - dm-tree==0.1.7
    - fastapi==0.80.0
    - filelock==3.7.1
    - fonttools==4.34.4
    - fsspec==2022.7.1
    - google-api-core==2.8.2
    - googleapis-common-protos==1.56.4
    - gpustat==1.0.0rc1
    - grpcio==1.43.0
    - gym==0.21.0
    - h11==0.13.0
    - imageio==2.21.0
    - iso8601==1.0.2
    - jsonschema==4.9.1
    - kiwisolver==1.4.4
    - kopf==1.35.6
    - kubernetes==24.2.0
    - lz4==4.0.2
    - matplotlib==3.5.2
    - msgpack==1.0.4
    - networkx==2.8.5
    - numpy==1.23.1
    - nvidia-ml-py==11.495.46
    - opencensus==0.11.0
    - opencensus-context==0.1.3
    - opentelemetry-api==1.1.0
    - opentelemetry-exporter-otlp==1.1.0
    - opentelemetry-exporter-otlp-proto-grpc==1.1.0
    - opentelemetry-proto==1.1.0
    - opentelemetry-sdk==1.1.0
    - opentelemetry-semantic-conventions==0.20b0
    - packaging==21.3
    - pandas==1.4.3
    - pillow==9.2.0
    - platformdirs==2.5.2
    - prometheus-client==0.13.1
    - protobuf==3.20.1
    - psutil==5.9.1
    - py-spy==0.3.12
    - pyarrow==6.0.1
    - pydantic==1.9.2
    - pyparsing==3.0.9
    - pyrsistent==0.18.1
    - python-json-logger==2.0.4
    - pytz==2022.1
    - pywavelets==1.3.0
    - pyyaml==6.0
    - ray==2.0.0
    - ray-cpp==2.0.0
    - scikit-image==0.19.3
    - scipy==1.9.0
    - smart-open==6.1.0
    - sniffio==1.2.0
    - starlette==0.19.1
    - tabulate==0.8.10
    - tensorboardx==2.5.1
    - tifffile==2022.8.3
    - uvicorn==0.16.0
    - virtualenv==20.16.2
    - websocket-client==1.4.0

```

### Reproduction script

```
import ray
import numpy as np
from ray.rllib.examples.env.bandit_envs_recommender_system import *
from ray.rllib.agents.ppo import PPOTrainer
config = {
"env": "ray.rllib.examples.env.bandit_envs_recommender_system.ParametricRecSys",

"framework": "tf",
"metrics_num_episodes_for_smoothing": 1000,

"env_config": {
        # Number of different categories a doc can have and a user can
        # have a preference for.
    #"num_categories": 2,
    # Number of docs to choose (a slate) from each timestep.
    "num_docs_to_select_from": 10,
    # Slate size.
    "slate_size": 1,
    # Re-sample docs each timesteps.
    "num_docs_in_db": 100,
    # Re-sample user each episode.
    "num_users_in_db": 1,
    # User time budget (determines lengths of episodes).
    "user_time_budget": 60.0
},
"model":
    {"fcnet_hiddens": [256, 256]},
"horizon": 10,
# Evaluation settings.
"evaluation_interval": 1,
"evaluation_num_workers": 0,
"evaluation_duration": 100,
"evaluation_duration_unit": "episodes",
"evaluation_parallel_to_training": False
}

trainer = PPOTrainer(config=config)
trainer.evaluate()
```

### Issue Severity

High: It blocks me from completing my task.

**Comment by ArturNiederfahrenhorst** [comment]
Thanks for raising this. I can reproduce.

**Comment by Rohan138** [comment]
Closing, fixed by #28358
